include/fiction/io/read_gate_level_layout.hpp

```diff
@@ -0,0 +1,771 @@
+                        {
+                            incoming_signal.x = std::stoull(incoming_signal_x->GetText());
+                        }
+                        else
+                        {
+                            throw gate_level_parsing_error("Error parsing gate_level file: no element 'x' in 'signal'");
+                        }
+
+                        // get y-coordinate of incoming signal
+                        auto* const incoming_signal_y = signal->FirstChildElement("y");
+                        if (incoming_signal_y != nullptr and incoming_signal_y->GetText())
+                        {
+                            incoming_signal.y = std::stoull(incoming_signal_y->GetText());
+                        }
+                        else
+                        {
+                            throw gate_level_parsing_error("Error parsing gate_level file: no element 'y' in 'signal'");
+                        }
+
+                        // get z-coordinate of incoming signal
+                        auto* const incoming_signal_z = signal->FirstChildElement("z");
+                        if (incoming_signal_z != nullptr and incoming_signal_z->GetText())
+                        {
+                            incoming_signal.z = std::stoull(incoming_signal_z->GetText());
+                        }
+                        else
+                        {
+                            throw gate_level_parsing_error("Error parsing gate_level file: no element 'z' in 'signal'");
+                        }
+
+                        gate.incoming.push_back(incoming_signal);
+                    }
+                }
+
+                gates.push_back(gate);
+            }
+
+            // sort gates ascending based on id
+            std::sort(gates.begin(), gates.end(), gate_storage::compare_by_id);
+
+            for (const auto& gate : gates)
+            {
+                const tile<Lyt> location{gate.loc.x, gate.loc.y, gate.loc.z};
+
+                if (gate.incoming.size() == 0)
+                {
+                    if (gate.type == "PI")
+                    {
+                        lyt.create_pi(gate.name, location);
+                    }
+                    else
+                    {
+                        throw gate_level_parsing_error("Error parsing gate_level file: unknown gate type: " +
+                                                       std::string(gate.type));
+                    }
+                }
+
+                else if (gate.incoming.size() == 1)
+                {
+                    const tile<Lyt> incoming_tile{gate.incoming.front().x, gate.incoming.front().y,
+                                                  gate.incoming.front().z};
+                    const auto      incoming_signal = lyt.make_signal(lyt.get_node(incoming_tile));
+
+                    if (gate.type == "PO")
+                    {
+                        lyt.create_po(incoming_signal, gate.name, location);
```

Comment:
You could wrap all `create_X` functions in another `if constexpr (has_create_X<Lyt>)` block to make sure the layout supports this designated type. And if not, default to the creation via truth table. Just an idea.